designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,359 @@
+
+
+# API Design
+## Approach A: Extending the Budget API to specify a reason 
+This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).
+
+### List Approach: Multiple Reasons per Budget - Recommended
+This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
+#### Proposed Spec
+Add a simple field "reasons" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // Reasons is a list of disruption reasons. If Reasons is not set, this budget applies to all methods.
+      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
+      // this budget will apply to all of them. If a reason is unspecified we will take the min value of this budget and the rest of the active budgets.
+      // if an unspecified reason exists we will also override all other reasons with its value if they are smaller than the unspecified reason.
+      // allowed reasons are "underutilized", "expired", "empty", "drifted"
+      // +kubebuilder:validation:MaxItems=5
+      // +kubebuilder:validation:Enum:={"underutilized","expired","empty","drifted"}
+      // +optional
+      Reasons []string `json:"reason,omitempty" hash:"ignore"`
+      // Nodes dictates the maximum number of NodeClaims owned by this NodePool
+      // that can be terminating at once. This is calculated by counting nodes that
+      // have a deletion timestamp set, or are actively being deleted by Karpenter.
+      // This field is required when specifying a budget.
+      // This cannot be of type intstr.IntOrString since kubebuilder doesn't support pattern
+      // checking for int nodes for IntOrString nodes.
+      // Ref: https://github.com/kubernetes-sigs/controller-tools/blob/55efe4be40394a288216dab63156b0a64fb82929/pkg/crd/markers/validation.go#L379-L388
+      // +kubebuilder:validation:Pattern:="^((100|[0-9]{1,2})%|[0-9]+)$"
+      // +kubebuilder:default:="10%"
+      Nodes string `json:"nodes" hash:"ignore"`
+      // Schedule specifies when a budget begins being active, following
+      // the upstream cronjob syntax. If omitted, the budget is always active. // Timezones are not supported. // This field is required if Duration is set. // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$` // +optional Schedule *string `json:"schedule,omitempty" hash:"ignore"` // Duration determines how long a Budget is active since each Schedule hit. // Only minutes and hours are accepted, as cron does not work in seconds.
```

Comment:
nit: fix the new lines on this
```suggestion
      // the upstream cronjob syntax. If omitted, the budget is always active. 
      // Timezones are not supported. This field is required if Duration is set. 
      // +kubebuilder:validation:Pattern:=`^(@(annually|yearly|monthly|weekly|daily|midnight|hourly))|((.+)\s(.+)\s(.+)\s(.+)\s(.+))$` 
      // +optional Schedule *string `json:"schedule,omitempty" hash:"ignore"` 
      // Duration determines how long a Budget is active since each Schedule hit. 
      // Only minutes and hours are accepted, as cron does not work in seconds.
```